### PR TITLE
Add a config property for the max number of rows to insert in SQL statements

### PR DESCRIPTION
**Pull request by flop01**
Add a "DB.InsertBatchMaxSize" with a default of 1000 to the node properties.

This property is used in two places known to generate very big SQL statements that exceed the maximum of certain DBMS, especially SQLite.

Fixes signum-network/signum-node#824

### CODE REVIEW

File: src/brs/db/sql/SqlAccountStore.java [hunk 1]
```diff
@@ -186,7 +186,7 @@ protected void bulkInsert(DSLContext ctx, Collection<Account.Balance> accounts)
               balance.getBalanceNqt(), balance.getUnconfirmedBalanceNqt(),
               balance.getForgedBalanceNqt(), true));
 
-          if(rows.size() >= 250000){
+          if(rows.size() >= Signum.getPropertyService().getInt(Props.DB_INSERT_BATCH_MAX_SIZE)){
```
Comment (ohager) [inline src/brs/db/sql/SqlAccountStore.java:189]: nice finding!

File: src/brs/props/Props.java [hunk 2]
```diff
@@ -128,6 +128,8 @@ public class Props {
   public static final Prop<Integer> BRS_BLOCK_CACHE_MB = new Prop<>("node.blockCacheMB", 40);
   public static final Prop<Integer> BRS_AT_PROCESSOR_CACHE_BLOCK_COUNT = new Prop<>("node.atProcessorCacheBlockCount", 1000);
 
+  public static final Prop<Integer> DB_INSERT_BATCH_MAX_SIZE = new Prop<>("DB.InsertBatchMaxSize", 1000);
```
Comment (ohager) [inline src/brs/props/Props.java:131]: Why 1000, and not e.g. 10000?

**Review by ohager — CHANGES_REQUESTED** [review]
Nice finding!

I just wonder why limit to 1.000 and not 10.000 as the issue is with value > 19.000 on sqlite?

In the case of account balance update the 250k batch size is in respect to Distribution of Token Holders, which can go up to 1.2M updates -> Having to update in 1k batches causes 1.2k iterations (and DB interactions) per distribution, while with 10k it would be 1/10th only - thus, expect in significantly better performance.

File: src/brs/props/Props.java (hunk 2 above)
Comment (frankTheTank72) [reply src/brs/props/Props.java:131]: I run it with 2500 and it worked well .. maybe 10k is already a problem for SQLite .. so as it is a parameter i would set it lower and if we get the 1.2M tx more often we can inform user to adjust the parameter ... so i would give with 2.500 as default - not higher

File: conf/node-default.properties [hunk 3]
```diff
@@ -326,6 +329,13 @@
 
 # DB.LockTimeout = 60
 
+## Maximum number of rows to be inserted in a single SQL statement.
+## Defaults to 1000, should be ok in most situations.
+## May be fine tuned according to your DBMS or to your machine performances.
+## Warning: a high value (> 19000 rows) is known to generate queries too big for an SQLite backend
```
Comment (flop01) [inline conf/node-default.properties:335]: this can be lowered to "> 15000 rows" : 15000 worked, 17000 triggered the bug.

File: conf/node-default.properties [hunk 4]
```diff
@@ -278,12 +278,15 @@
...
 
 ## Your keystore file and password, required if uiSSL or apiSSL are enabled.
+
 # API.SSL_keyStorePath     = keystore
 # API.SSL_keyStorePassword = password
 
 ## If you use https://certbot.eff.org/ to issue your certificate, provide below the path for your keys.
 ## BRS will automatically create the keystore file using the password above and will reload it weekly.
 ## Make sure you configure certbot to renew your certificate automatically so you don't need to worry about it.
+## Note, you need 'openssl' on your path for this to work, most Linux distributions have it already.
```
Comment (flop01) [inline conf/node-default.properties:288]: dunno why this change, can be discarded, it was not intentional from me

File: src/brs/props/Props.java (hunk 2 above)
Comment (ipr0310) [reply src/brs/props/Props.java:131]: A big factor on this number could also be when syncing a node with a HDD or SSD

Comment (frankTheTank72) [reply src/brs/props/Props.java:131]: I assume it is a default value which may be set to 1M bytes or so by SQLite and if the statement goes above it triggers the error.. 

Maximum Length Of An SQL Statement

The maximum number of bytes in the text of an SQL statement is limited to SQLITE_MAX_SQL_LENGTH which defaults to 1,000,000,000.


Found on https://www.sqlite.org/limits.html

--

Maybe 5k is a good value in the middle

**Comment by flop01** [comment]
Here is more food for thought regarding batches sizes.
I've made a few figures regarding the size of the insertions, from genesis block up to block 

Here's what I found for the two types of transactions that are know to generate a big number of database insertions in a single block :

# account balances changes transactions
key figures:
- there's a total of 1347377 blocks since genesis as of this post
- over this grand total, there are only 18 blocks that would make more than 10000 insertions at once
- and only 259 blocks that would make more than 1000 insertions

Here is what it looks like on a chart.
I cut the Y axis at 10000 so it looks better. There are of course much more than 10000 blocks that resulted in a number of rows inserted in the low range like 1, 2, 3, 4... rows.

Y is amount of blocks since genesis
Screenshot from 2024-11-26 08-34-54
X is amount of insertions in block

# indirect incoming transactions

There is a total of 533857 indirect transactions that result in row insertions in the database
The highest number of rows inserted at once was 557 !

Here's the same chart as above, for indirect transactions :
Screenshot from 2024-11-26 08-40-00

# conclusion
I'd say it doesn't really matter to choose between a maximum batch size of 1000, 2500 or 10000 for database insertions, since there are actually very very few big database insertions in the whole 10 years of blockchain history, with only 259 blocks resulting in a more than 1000 rows insertion at once.

I also think staying at 1000 rows max might play better with computers that don't have a very big amount of RAM.
I asked grok about sqlite, postgres and mariadb, and the reply hinted that mariadb would be the bottleneck here, with a suggestion of no more than 1000 to 5000 rows at once:
Screenshot from 2024-11-26 08-54-57

So in the end, I'd probably leave a default of 1000 rows maximum for this new setting, or maybe bump it to 2500 maximum.

What do you think ?

File: src/brs/props/Props.java (hunk 2 above)
Comment (flop01) [reply src/brs/props/Props.java:131]: > A big factor on this number could also be when syncing a node with a HDD or SSD The specs of the `server/computer`
Also why I thought 1000 should work for most hardware.

See also my comment below with statistics.
Only 259 blocks are more than 1000 rows to insert, since genesis block !

1000 or 2500 would be fine, I think.

File: conf/node-default.properties (hunk 3 above)
Comment (flop01) [reply conf/node-default.properties]: edited this in a new commit in this PR. Now warns about values above 15000 rows

File: conf/node-default.properties (hunk 4 above)
Comment (frankTheTank72) [reply conf/node-default.properties:288]: so remove it and update the commit

File: src/brs/props/Props.java (hunk 2 above)
Comment (ohager) [reply src/brs/props/Props.java:131]: :+1:

File: src/brs/db/sql/SqlIndirectIncomingStore.java [hunk 5]
```diff
@@ -60,7 +62,7 @@ void save(DSLContext ctx, Collection<IndirectIncoming> indirectIncomings) {
         while (iterator.hasNext()) {
           List<Record5<Long, Long, Long, Long, Integer>> rows = new ArrayList<>();
           // break into batches
-          for (int i = 0; i < 250000 && iterator.hasNext(); i++) {
+          for (int i = 0; i < Signum.getPropertyService().getInt(Props.DB_INSERT_BATCH_MAX_SIZE) && iterator.hasNext(); i++) {
```
Comment (ohager) [inline src/brs/db/sql/SqlIndirectIncomingStore.java:65]: btw: this should go out of the loops!

File: src/brs/db/sql/SqlAccountStore.java (hunk 1 above)
Comment (ohager) [reply src/brs/db/sql/SqlAccountStore.java:189]: should go outside the loop, and ideally maxed to 10k

**Review by ohager — APPROVED** [review]
Good catch... 
Ideally, the property value is not being used inside the loops and due to risk of execution failure maxed to lets say 10k or 15k

File: src/brs/props/Props.java [hunk 6]
```diff
@@ -128,7 +128,7 @@ public class Props {
   public static final Prop<Integer> BRS_BLOCK_CACHE_MB = new Prop<>("node.blockCacheMB", 40);
   public static final Prop<Integer> BRS_AT_PROCESSOR_CACHE_BLOCK_COUNT = new Prop<>("node.atProcessorCacheBlockCount", 1000);
 
-  public static final Prop<Integer> DB_INSERT_BATCH_MAX_SIZE = new Prop<>("DB.InsertBatchMaxSize", 1000);
+  public static final Prop<Integer> DB_INSERT_BATCH_MAX_SIZE = new Prop<>("DB.InsertBatchMaxSize", 10000);
```
Comment (flop01) [inline src/brs/props/Props.java:131]: as requested, 10000 as default value

**Review by frankTheTank72 — APPROVED** [review]
